snooty/util.py

```diff
@@ -60,10 +61,17 @@ def reroot_path(
     return rel_fn, project_root.joinpath(rel_fn).resolve()
 
 
-def get_files(root: PurePath, extensions: Container[str]) -> Iterator[Path]:
+def get_files(root: Path, extensions: Container[str]) -> Iterator[Path]:
     """Recursively iterate over files underneath the given root, yielding
-    only filenames with the given extensions."""
-    for base, dirs, files in os.walk(root):
+    only filenames with the given extensions. Symlinks are followed, but
+    any given concrete directory is only scanned once."""
+    seen: Set[Path] = set()
+
+    for base, dirs, files in os.walk(root, followlinks=True):
+        dirs_set = set(Path(d).resolve() for d in dirs)
+        dirs[:] = [d.name for d in (dirs_set - seen)]
```

Comment:
What is this new value of `dirs` being used for?

Reply:
Excellent question! This is an oddball eccentricity of the `os.walk()` interface: the second member of the return tuple, `dirs`, lists the subdirectories that will be explored next. But if you modify this list, you get to filter which subdirectories you actually want to explore!

Note the comments on the `topdown` argument in https://docs.python.org/3/library/os.html#os.walk

Reply:
This makes a lot more sense now. Thank you!